### DISCUSSION TITLE
Getting started w/ Ryven

### DISCUSSION THREAD

**Discussion by WillAdams** [Tanneguydv/Pythonocc-nodes-for-Ryven#13] [Q&A]
Okay, here are some notes/questions on this (Note that the original attempt was to install into Ubuntu, but later switched to Windows) and what I hope are workable step-by-step instructions:

First, it is important that one have a clean/consistent Python environment --- uninstall any environments which are not being used. If need be, remove all and start fresh.

Second, it seems necessary to place PythonOCC in a specific environment --- attempts to install it globally and then use it w/o specifying an environment were unsuccessful. (This installation technique is noted on the github page for PythonOCC). Once one has created and activated the environment (named pyoccenv in the instructions below) each successive step must be done w/ that environment active.

Third, note that several different install methodologies/commands are used because this tool follows the Unix philosophy of using the appropriate small tool for any given task.

Lastly, note that the tool has several layers each of which has to be set in place before the one on top of it may be made use of, and that as noted above, a critical glue holding things together is the named environment which must be active once it is created.

 - used a download of Python from python.org and installed it per: https://docs.python.org/3/using/windows.html
 - Install miniconda w/ the defaults: https://docs.conda.io/en/latest/miniconda.html
 - install PythonOCC (note that using python=3.10 below did not work):

    conda create --name=pyoccenv python=3.9
    conda activate pyoccenv
    conda install -c conda-forge pythonocc-core=7.5.1 occt=7.5.1

 - install PythonOCC-utils: Download .zip archive of: https://github.com/tpaviot/pythonocc-utils (or git clone) and unzip to:

    C:\Users\<YOUR USER DIRECTORY>\miniconda3\envs\pyoccenv

Then, cd into that directory:

    cd C:\Users\<YOUR USER DIRECTORY>\miniconda3\envs\pyoccenv\pythonocc-utils-master

    python setup.py install

 - install Ryven:
 
    pip install ryven

 - download and unzip: https://github.com/Tanneguydv/Pythonocc-nodes-for-Ryven/archive/refs/heads/main.zip somewhere convenient
 - also download any of the demo .json files which are of interest

At this point, it should work to launch ryven when the pyoccenv is active. 

If you have a saved .json file which you wish to load, click on "load example" when you first launch Ryven.

When launching Ryven (or loading an example) and making a new project File | Import Example Nodes --- this should open in ~/.local/lib/python3.9/site-packages/ryven/example_nodes --- go into the std directory and select nodes.py there --- it will also be necessary to load the nodes.py file for PythonOCC which is in the unzipped copy of Pythonocc-nodes-for-Ryven:

![image](https://user-images.githubusercontent.com/3543154/152267597-9649026d-c2be-4863-b0b1-47f5c00638aa.png)

Notes:

The program Ryven doesn't warn about unsaved changes when you close it.

When you save a project the .json file extension isn't included automatically at the end of the filename --- add manually to open projects (this will be fixed in a future version).
 
It seems to be necessary to quit and then restart in order to load a new project.

Anyway success!

![image](https://user-images.githubusercontent.com/3543154/152267227-16dd6943-6ec4-4d98-9aae-1337aee67c0e.png)

**Comment by leon-thomm** [comment]
some comments:

> When launching Ryven and making a new project File | Import Example Nodes --- this should open in ~/.local/lib/python3.9/site-packages/ryven/example_nodes --- go into the std directory and select nodes.py there

Depends where your python environment is located that you installed Ryven in. For using PythonOCC, however, this path seems not correct, as it points to a Ryven installation of your system-wide Python3.9, but you should use PythonOCC in an Anaconda environment, which has its own Python path.

> Kind of mystified by whether I managed to get pythonocc-core installed, or how I'd test that. Also, what 3D viewer should be used to display a 3D result, and how do you link in PythonOCC when loading the file demo_example.json --- I'm just not seeing it.

The link in the README directs to https://anaconda.org/conda-forge/pythonocc-core which shows the command to install it in anaconda. If your anaconda environment is working and you installed pythonocc-core in it, you should see it in the list by running `conda list`.

In case you didn't use Anaconda previously check out how to set it up on your system, create a new conda environment, install all dependencies listed in the README of this repo into this environment, clone the repository and start ryven from within the anaconda environment.

**Comment by Tanneguydv** [comment]
If you're not familiar with python, anaconda and so on, on Windows OS an easy way to proceed is to :
1. Install Python (example 3.7) on your computer
2. download miniconda (instead of all Anaconda) : [https://docs.conda.io/en/latest/miniconda.html](https://docs.conda.io/en/latest/miniconda.html)
3. Run The Miniconda Prompt
4. Type those lines to create a virtual environment, called pyoccryv3env, using python 3.7, install pythonocc, and Ryven:
```
conda create --name=pyoccryv3env python=3.7
conda activate pyoccryv3env
conda install -c conda-forge pythonocc-core=7.5.1 occt=7.5.1
pip install ryven
```
On Windows your environment files will be here : `C:\Users\your_name\.conda\envs`
5. Clone this repo [https://github.com/tpaviot/pythonocc-utils](https://github.com/tpaviot/pythonocc-utils) here:
`C:\Users\your_name\.conda\envs\pyoccryv3env\Lib\site-packages`
6. create a `.bat` file (using a simple text editor) with this text :
```
call activate pyoccryv3env
ryven
pause
```
7. Double click on the` .bat` file to open Ryven (you can create a shortcut of this file with an icon too)
8. Clone pythonocc_nodes_for_ryven on your computer and then load the `nodes.py` file from Ryven

**Reply by WillAdams** [reply]
I don't believe .bat files run on Linux.

**Reply by leon-thomm** [reply]
they don't, but it's conceptually the same in .sh files, just have use `conda activate` or `source activate` or something like that, I think

**Reply by WillAdams** [reply]
Thanks. I'll try to look into it on that on that basis.

**Reply by WillAdams** [reply]
The thing I don't understand is why pip gets used for one thing, and anaconda for others.

I'm trying again w/ winpython, but I don't see how to make anaconda or miniconda install inside winpython.

Why can't all the installation be done using pip?

**Reply by WillAdams** [reply]
Turning this around. Uninstalled all Python installations and related utilities and installed Python 3.9 using Anaconda. conda install pip and pip install ryven failed (ryven wouldn't start) so trying conda install -c conda-forge ryven --- which is taking forever

and which finally ended with the error:

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

What would be involved in getting this all added to winpython?

**Reply by leon-thomm** [reply]
ryven is just a python package (pip), anaconda is an environment platform for python that is extremely widely used, much more than winpython. The ryven conda package is very new and there might be issues on some systems, but you can still install packages in an anaconda environment using pip:
```
$ conda install pip
```
make sure that you are using anaconda's python's pip
```
$ which pip
```
install and run ryven
```
$ pip install ryven
$ ryven
```
if that fails, it's gonna be hard for us to tell what's the problem without an error output

**Reply by WillAdams** [reply]
The error messages when using Conda were:

(base) C:\Users\willa>conda config --add channels conda-forge
Warning: 'conda-forge' already in 'channels' list, moving to the top

(base) C:\Users\willa>conda config --set channel_priority strict

(base) C:\Users\willa>conda install ryven
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: -
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
Examining @/win-64::__win==0=0:  25%|█████████████                                       | 1/4 [00:00<00:02Examining @/win-64::__win==0=0:  50%|██████████████████████████                          | 2/4Examining @/win-64::__archspec==1=x86_64:  50%|█████████████████████                     | 2/4 [00:Examining ryven:  75%|█████████████████████████████████████████████████failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions
Note that strict channel priority may have removed packages required for satisfiability.

Made in issue on that in what I hope was the right place.

**Reply by WillAdams** [reply]
Okay, I tried again, this time cleaning up more python installations and leftovers from Python installations than I knew I had (my apologies, didn't know I was channeling my inner Randall Munroe: https://xkcd.com/1987/ ) and this time the install seemed to work fine:

 - first I uninstalled and cleaned up after more Python installs than I could remember having installed
 - then I downloaded Python from python.org and installed it per: https://docs.python.org/3/using/windows.html
 - Install miniconda w/ the defaults: https://docs.conda.io/en/latest/miniconda.html
 - install PythonOCC: conda install -c conda-forge pythonocc-core
 - Go to: https://github.com/tpaviot/pythonocc-utils and download as a .zip archive and unzip
 - then run the setup.py command using the command: python setup.py install

initial nodes are loaded from:

C:\Users\<YOUR USER DIRECTORY>\AppData\Local\Programs\Python\Python310\Lib\site-packages\ryven\example_nodes\std

Download the .zip of: https://github.com/Tanneguydv/Pythonocc-nodes-for-Ryven and unzip it

Launch Ryven and import std from the afore-mentioned link, and PythonOCC from the copy of nodes.py from the above archive --- except it doesn't seem to work:

![image](https://user-images.githubusercontent.com/3543154/152079034-ca25508e-c2cb-4fec-9a73-d8e10b2d5e2a.png)

Shifting discussion to: https://github.com/Tanneguydv/Pythonocc-nodes-for-Ryven/issues/11 which I believe is the last hurdle.

**Reply by WillAdams** [reply]
Got this working, and edited my initial post so that hopefully it will be useful for other folks.

Please feel free to copy and use it to update any documentation.

**Reply by WillAdams** [reply]
Not seeing a noticeable difference in the 3 demo examples --- are they supposed to be the same? If they're subtly different, what is it that one should learn from them?
